## IESO

In [15]:
%load_ext autoreload
%autoreload 2

In [34]:
# # load from github
# import urllib.request
# api_url = 'https://raw.githubusercontent.com/tanmayyb/ele70_bv03/refs/heads/main/api/datasets.py'
# exec(urllib.request.urlopen(api_url).read())

# load from local
import sys
import os
project_root = os.path.abspath(os.path.join(os.getcwd(), '../..'))
sys.path.append(project_root)
from api.datasets import IESODataset

In [35]:
# dataset = IESODataset('fsa')
# dataset.load_dataset(download=False)
# dataset.selected_dates

dataset = IESODataset('fsa')
# dataset.load_dataset()
# dataset.selected_dates

Time range: 201801 to 202411


| Property | FSA Type | Zonal Type |
|----------|----------|------------|
| dataset_type | "fsa" | "zonal" |
| region | Only "ON" supported | Only "ON" supported |
| target_options | List of Ontario cities | Fixed list: ["Northwest", "Northeast", "Ottawa", "East", "Toronto", "Essa", "Bruce", "Southwest", "Niagara", "West", "Zone Total"] |
| target_val | List of postal codes for selected city | Same as selected zone name |
| filetype | "zip" | "csv" |
| date_type | Monthly (YYYYMM) | Yearly (YYYY) |
| Data Granularity | FSA (Forward Sortation Area) level consumption | Zonal level demand |
| Data Format | Total consumption by postal code | Demand by geographical zone |


| Method | FSA Type | Zonal Type |
|--------|----------|------------|
| set_target(target_idx: int) | Sets city and corresponding postal codes<br>Example: set_target(0) might select "Toronto" and its FSAs | Sets single zone name<br>Example: set_target(4) selects "Toronto" zone |
| load_dataset(start_date, end_date, target_idx, download=True) | start_date: YYYYMM format (e.g., 201801)<br>end_date: YYYYMM format<br>Example: load_dataset(201801, 201812, 0, download=True) | start_date: YYYY format (e.g., 2018)<br>end_date: YYYY format<br>Example: load_dataset(2018, 2018, 4, download=True) |
| get_target_options() | Returns list of available cities | Returns list of 11 zones:<br>["Northwest", "Northeast", "Ottawa", etc.] |
| parse_dataset(chunk_size=4) | Processes FSA consumption data<br>Groups by DATE, HOUR for selected postal codes | Processes zonal demand data<br>Direct selection of zone column |
| save_dataset(filepath=None) | Saves data with FSA metadata:<br>- City name<br>- Postal codes<br>- Monthly files | Saves data with zonal metadata:<br>- Zone name<br>- Yearly files |
| load_from_json(filepath=None) | Loads FSA dataset with:<br>- Consumption by postal code<br>- DateTime in monthly granularity | Loads zonal dataset with:<br>- Demand by zone<br>- DateTime in yearly granularity |

In [50]:
dataset.load_dataset(start_date=202001, end_date=202101)


## Climate

In [ ]:
# class WeatherDataset(Dataset):
#   def __init__(self, region: str = "ON", sampling_seed: int = 42, sample_num: int = 10):
#     super().__init__(region, "CANADA")
#     self.dataset_type = "WEATHER"
    
#     # add zonal list
#     # add sampling based on zonal list

In [ ]:
# class DatasetLoader():
#   def __init__(self, dataset: Dataset):
#     self.dataset = dataset

#   def load_dataset(self, dataset, start_date: int = None, end_date: int = None, download: bool = False):
#     if start_date > end_date:
#       raise ValueError("Start date must be before end date")

#     files = dataset.files
#     file_type = dataset.filetype
#     dates = dataset.dates
#     date_type = dataset.date_type

#     if download:
#       self.dataset.download_dataset()
#     else:
#       self.dataset.load_dataset_from_local()

  # def download_dataset(self, dataset, file_type, date_type, start_date: int, end_date: int):
  # # make sure data folder exists
  #   if file_type == "zip":
  #     self.download_zip_dataset(files, dates, date_type)
  #   elif file_type == "csv":
  #     self.download_csv_dataset(files, dates, date_type)

  # def download_zip_dataset(self, files, dates, date_type):
  #   for file in files:
  #     url = file